google_guest_agent/command/command.go

```diff
@@ -0,0 +1,140 @@
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+// Package command facilitates calling commands within the guest-agent.
+package command
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/GoogleCloudPlatform/guest-agent/google_guest_agent/cfg"
+)
+
+// Handler functions are the business logic of commands. They must process json
+// encoded as a byte slice which contains a Command field and optional arbitrary
+// data, and return json which contains a Status, StatusMessage, and optional
+// arbitrary data (again encoded as a byte slice). Returned errors will be
+// passed onto the command requester.
+type Handler func([]byte) ([]byte, error)
+
+// Request is the basic request structure. Command determines which handler the
+// request is routed to. Callers may set additional arbitrary fields.
+type Request struct {
+	Command string
+}
+
+// Response is the basic response structure. Handlers may set additional
+// arbitrary fields.
+type Response struct {
+	// Status code for the request. Meaning is defined by the caller, but
+	// conventially zero is success.
+	Status int
+	// StatusMessage is an optional message defined by the caller. Should generally
+	// help a human understand what happened.
+	StatusMessage string
+}
+
+var (
+	// CmdNotFoundError is return when there is no handler for the request command
+	CmdNotFoundError = Response{
+		Status:        101,
+		StatusMessage: "Could not find a handler for the requested command",
+	}
+	// BadRequestError is returned for invalid or unparseable JSON
+	BadRequestError = Response{
+		Status:        102,
+		StatusMessage: "Could not parse valid JSON from request",
+	}
+	// ConnError is returned for errors from the underlying communication protocol
+	ConnError = Response{
+		Status:        103,
+		StatusMessage: "Connection error",
+	}
+	// TimeoutError is returned when the timeout period elapses before valid JSON is receieved
+	TimeoutError = Response{
+		Status:        104,
+		StatusMessage: "Connection timeout before reading valid request",
+	}
+	// HandlerError is returned when the handler function returns an non-nil error. The status message will be replaced with the returnd error string.
+	HandlerError = Response{
+		Status:        105,
+		StatusMessage: "The command handler encountered an error processing your request",
+	}
+	// InternalErrorCode is the error code for internal command server errors. Returned when failing to marshal a response.
+	InternalErrorCode = 106
+	internalError     = []byte(`{"Status":106,"StatusMessage":"The command server encountered an internal error trying to respond to your request"}`)
+	handlersMu        = new(sync.RWMutex)
```

Comment:
Can we wrap these `handlersMu` and `handlers` in a instance object?

Reply:
I don't see a way to implement this cleanly in the Register and Unregister calls, since cmdServer is nil if handlers register before it's initialized and we don't want callers besides the main agent calling `Init()`
```
// With a feature flag
if !cfg.Get().Unstable.CommandMonitorEnabled {
  return fmt.Errorf("disabled")
}
for cmdServer == nil {
}
cmdServer.handlers[cmd] = handler
```
But the infinite loop until the server exists strikes me as a bad pattern.

Reply:
We are exposing `Server` right? We should protect `Server` as server is an implementation detail and expose the command monitor. So you can have:

```
type Monitor struct {
   server Server
   handlersMu sync.RWMutex
   handlers map[string]Handler
}
```
Note: you could un-export server though since we want to hide it.

then you would have a:

```
func Init(...) Monitor {
    return &Monitor{server: initServer(), ....}
}
```
Where `initServer()` would be your current `Init()` implementation.

Or either, to be consistent with other packages we don't return `Monitor` but initialize the singleton variable and use a `Get()` to return the instance.

With that the calling pattern would look like this:

```
if err := command.Init(); err != nil {
    log.Error(err)
}
defer command.Close()
....

command.Get().Subscribe()
...
```

`command.Close()` would use the singleton variable.

PS: if we have a singleton `Monitor` we don't need the singleton `Server`.

Reply:
ACK. I misunderstood what you were saying at first, I see now.